Tools/Scripts/hooks/pre-push

```diff
@@ -0,0 +1,330 @@
+MAX_LEVEL = max(SECURITY_LEVELS.values())
+
+
+def print_error(why):
+    if sys.stdout.isatty():
+        RED_TEXT_FORMAT = '\033[31m{}\033[0m\n'
+        sys.stderr.write(RED_TEXT_FORMAT.format(why))
+    else:
+        sys.stderr.write('{}\n'.format(why))
+
+
+DEFAULT_MODE = 'default' # Forbid class 1, prompt for class 2 and 3
+PUBLISH_MODE = 'publish' # Prompt for class 1, forbid class 2 and 3
+NO_RADAR_MODE = 'no-radar' # Forbid class 1 and 2, don't compute class 3
+MODE = os.environ.get('PUSH_HOOK_MODE', {{ default_pre_push_mode }})
+
+
+if MODE not in (DEFAULT_MODE, PUBLISH_MODE, NO_RADAR_MODE):
+    print_error("'{}' is not a recognized push mode".format(MODE))
+    sys.exit(1)
+
+
+REPOSITORY = None
+try:
+    import webkitpy
+    from webkitcorepy import Version
+    from webkitscmpy import local, Commit
+    from webkitbugspy import version
+
+    if MODE != NO_RADAR_MODE and version >= Version(0, 9, 8):
+        REPOSITORY = local.Git(os.getcwd())
+
+except:
+    pass
+
+if not REPOSITORY:
+    print_error('Checkout too out-of-date to check bug status of novel commits')
+
+
+def security_level_for_remote(name, remote):
+    match = REMOTE_RE.match(remote)
+    if not match:
+        return SECURITY_LEVELS.get(name, None)
+    user, repository = match.group('path').split('/', 1)
+    split_repository = repository.split('-', 1)
+    candidates = [
+        '{}:{}'.format(match.group('host'), match.group('path')),
+        '{}:{}/{}'.format(match.group('host'), repository, repository),
+    ]
+    if '-' in repository:
+        repository, team = repository.split('-', 1)
+        candidates.append('{}:{}/{}'.format(match.group('host'), team, repository))
+        candidates.append('{}:{}/{}-{}'.format(match.group('host'), repository, repository, team))
+
+    for candidate in candidates:
+        if SECURITY_LEVELS.get(candidate, None) is not None:
+            return SECURITY_LEVELS[candidate]
+    return SECURITY_LEVELS.get(name, None)
+
+
+def security_level_for_remotes(remotes, name_to_remote=None):
+    if not name_to_remote:
+        _, name_to_remote = remote_name_mappings()
+    security_levels = [security_level_for_remote(name, name_to_remote.get(name, name)) for name in remotes]
+    if 0 in security_levels:
+        return 0
+    if None in security_levels:
+        return None
+    return min(security_levels)
+
+
+def commits_in(from_ref, to_ref):
+    command = ['git', 'rev-list']
+    if from_ref:
+        command.append('{}..{}'.format(from_ref, to_ref))
+    else:
+        command.append(to_ref)
+    try:
+        proc = subprocess.Popen(
+            command,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            **ENCODING_KWARGS
+        )
+        line = proc.stdout.readline()
+        while line:
+            yield line.rstrip()[:TRUNCATED_COMMIT_LEN]
+            line = proc.stdout.readline()
+    finally:
+        if proc and proc.poll() is None:
+            proc.kill()
```

Comment:
The trouble with `check_output` here is that if we're publishing a _new_ branch, we won't have a base commit, so we'll have to grab _all_ commits. My thinking was that we could start reasoning about the first handful of commits before `git rev-list` returned, and if the 2nd or 3rd commit already existed on a remote (which is a pretty common case, I would expect, since that's basically what happens when pushing a novel PR branch). Let me do some testing to see how bad `git rev-list` is with everything, and if we can get early-exit to work with generators.

Reply:
I double-checked this, the `finally` block gets rid of the zombie process, this little test program can verify that:

```
#!/usr/bin/env python3

import os
import subprocess
import sys
import time


repo = os.path.join(os.path.abspath(os.path.dirname(__file__)), 'OpenSource')

def log():
    command = ['git', 'log', 'HEAD', '--oneline']
    proc = None
    try:
        proc = subprocess.Popen(
            command, cwd=repo,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            encoding='utf-8'
        )
        line = proc.stdout.readline()
        while line:
            yield line.rstrip()
            line = proc.stdout.readline()
    finally:
        if proc and proc.poll() is None:
            proc.kill()

cnt = 0
for line in log():
    if cnt > 10:
        break
    print(line)
    cnt += 1

time.sleep(60)

```

Reply:
Fascinating. I guess it runs when the generator goes out of scope?